modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala

```diff
@@ -16,47 +16,91 @@
-      PayloadPath.root,
-      "decoding is unsupported, encoding only",
-      "TODO: integrate with XML decoders"
-    )
-  )
-
-  override type Codec[A] = AwsQueryCodec[A]
-  override type Cache = CompilationCache[AwsQueryCodec]
 
-  override def createCache(): Cache = CompilationCache.make[Codec]
+  override type Codec[A] = Either[AwsQueryCodec[A], XmlDecoder[A]]
+  override type Cache = CompilationCache[AwsQueryCodec]
 
-  override def mediaType[A](codec: Codec[A]): HttpMediaType =
-    HttpMediaType("application/x-www-form-urlencoded")
+  override def createCache(): Cache = CompilationCache.make[AwsQueryCodec]
 
   override def compileCodec[A](
       schema: Schema[A],
       cache: Cache
-  ): Codec[A] = schema.compile(new AwsSchemaVisitorAwsQueryCodec(cache))
+  ): Codec[A] =
+    schema.hints match {
+      case InputOutput.hint(InputOutput.Input) =>
+        Left(schema.compile(new AwsSchemaVisitorAwsQueryCodec(cache)))
+      case InputOutput.hint(InputOutput.Output) =>
+        Right(schema.compile(XmlDecoderSchemaVisitor))
+    }
+
+  override def mediaType[A](codec: Codec[A]): HttpMediaType =
+    HttpMediaType("application/x-www-form-urlencoded")
 
   override def decodeFromByteArrayPartial[A](
       codec: Codec[A],
       bytes: Array[Byte]
-  ): Either[PayloadError, BodyPartial[A]] = decodingError
+  ): Either[PayloadError, BodyPartial[A]] = codec match {
+    case Left(_) =>
+      Left(
+        PayloadError(
+          PayloadPath.root,
+          "",
+          "Invalid codec: got AWS Query encoder, expected XML decoder"
+        )
+      )
+    case Right(xmlDecoder) =>
+      Stream
+        .emit[SyncIO, String](new String(bytes, "UTF-8"))
+        .through(events[SyncIO, String]())
+        .through(documents[SyncIO, XmlDocument])
+        .map(doc => XmlCursor.fromDocument(doc))
+        .map(cursor => xmlDecoder.decode(cursor))
+        .rethrow
+        .head
+        .compile
+        .lastOrError
+        .map(a =>
+          // TODO: have no idea if it's correct or not
+          BodyPartial(Map.empty.withDefaultValue(a))
+        )
+        .attempt
+        .unsafeRunSync()
+        .leftMap(e =>
+          PayloadError(
+            PayloadPath.root,
+            "",
+            s"Failed to decode the response, message: ${e.getMessage}"
+          )
+        )
+  }
 
   override def decodeFromByteBufferPartial[A](
       codec: Codec[A],
       bytes: ByteBuffer
-  ): Either[PayloadError, BodyPartial[A]] = decodingError
+  ): Either[PayloadError, BodyPartial[A]] =
+    decodeFromByteArrayPartial(codec, bytes.array())
 
-  override def writeToArray[A](
-      codec: Codec[A],
-      value: A
-  ): Array[Byte] =
-    codec(value).render.getBytes("UTF-8")
+  override def writeToArray[A](codec: Codec[A], value: A): Array[Byte] =
+    codec match {
+      case Left(encoder) => encoder(value).render.getBytes("UTF-8")
+      case Right(_) =>
+        throw new RuntimeException("Boom! TODO: define a proper error")
```

Comment:
Is it safe to simply throw out an error over here?

Reply:
yeah.

Reply:
Done